mlir/lib/Dialect/SCF/IR/SCF.cpp

```diff
@@ -1673,7 +1674,12 @@ struct ForallOpIterArgsFolder : public OpRewritePattern<ForallOp> {
     for (OpResult result : forallOp.getResults()) {
       OpOperand *opOperand = forallOp.getTiedOpOperand(result);
       BlockArgument blockArg = forallOp.getTiedBlockArgument(opOperand);
-      if (result.use_empty() || forallOp.getCombiningOps(blockArg).empty()) {
+      SmallVector<Operation *> combiningOps =
+          forallOp.getCombiningOps(blockArg);
+      if ((result.use_empty() &&
+           llvm::all_of(combiningOps,
+                        [](Operation *op) { return op->use_empty(); })) ||
+          combiningOps.empty()) {
```

Comment:
```suggestion
      if ((result.use_empty() &&
           llvm::all_of(forallOp.getCombiningOps(blockArg),
                        [](Operation *op) { return op->use_empty(); }))) {
```

The `empty()` check is redundant: `all_of` returns `true` on empty ranges.

Also which test is covering this change?

Reply:
Actually such change is not correct when combiningOps is empty.

The test to test it is [here](https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/SCF/canonicalize.mlir#L1828).

Reply:
my bad. this is reverted.